draft-ietf-mls-protocol.md

```diff
@@ -1435,48 +1461,219 @@ all arrive at the following state:
-   group shared secret
+The Application AEAD keying material is generated from the following
+input values:
 
-Possible candidates for that are:
+- The Application Secret value;
+- A purpose value indicating the specific value being generated;
+- The length of the key being generated.
 
- * the key used for the previous message (hash ratcheting)
- * the counter of the previous message (needs to be known to new members of the group)
- * the hash of the previous message (proof that other participants saw the same history)
+Note, that because the identity of the participant using the keys to send data
+is included in the initial Application Secret, all successive updates to the
+Application secret will implicitly inherit this ownership.
 
-The requirement for this is that all participants know these values.
-If additional clear-text fields are attached to messages (like the counter), those
-fields MUST be protected by a signed message envelope.
+All the traffic keying material is recomputed whenever the underlying
+Application Secret changes.
+
+
+## Message Encryption and Decryption
+
+The Group participants MUST use the AEAD algorithm associated with
+the negotiated MLS ciphersuite to AEAD encrypt and decrypt their
+Application messages and sign them as follows:
+
+~~~~~
+    struct {
+        opaque content<0..2^32-1>;
+        opaque signature<0..2^16-1>;
+        uint8 zeros[length_of_padding];
+    } ApplicationPlaintext;
+
+    struct {
+        uint8  group[32];
+        uint32 epoch;
+        uint32 generation;
+        uint32 sender;
+        opaque encrypted_content<0..2^32-1>;
+    } Application;
+~~~~~
+
+The Group identifier and epoch allow a device to know which Group secrets
+should be used and from which Epoch secret to start computing other secrets
+and keys. The participant identifier is used to derive the participant
+Application secret chain from the initial shared Application secret.
+The application generation field is used to determine which Application
+secret should be used from the chain to compute the correct AEAD keys
+before performing decryption.
+
+The signature field allows strong authentication of messages:
+
+~~~
+    struct {
+        uint8  group[32];
+        uint32 epoch;
+        uint32 generation;
+        uint32 sender;
+        opaque content<0..2^32-1>;
+    } MLSSignatureContent;
+~~~
 
-Alternatively, the hash of the previous message can also be included as an additional
-field rather than change the encryption key. This allows for a more flexible approach,
-because the receiving party can choose to ignore it (if the value is not known, or if
-transcript security is not required).
+The signature used in the MLSPlaintext is computed over the MLSSignatureContent
+which covers the metadata information about the current state
+of the group (group identifier, epoch, generation and sender's Leaf index)
+to avoid Group participants to impersonate other participants and in order
```

Comment:
"to avoid..." -> "to prevent Group participants from impersonating"

Reply:
Fixed, Thanks !